### DISCUSSION TITLE
Difference in behavior of URLLoader and URLStream in Windows and MacOS

### DISCUSSION THREAD

**Discussion by sergeysoftech** [airsdk/Adobe-Runtime-Support#1856] [Q&A]
Hello,

Our code in some circumstances can generate a lot of parallel web requests. Such requests were generated by URLLoader and URLStream classes. Let's say there is about of 20 requests in one time, and user can make actions to generate next 20 requests right after these requests were proceeded. All requests go to the same server, difference only with parameters.
During testing we hit that these requests behaves very differently in case we run the code on Windows and MacOS:
Windows:
1) There are always only 4 active connections max. All others are in queue.
2) On completing the requests were closed (or reused). So there is always 4 or less active connections.
And such behavior is fine for us.

MacOS:
1) All requests are thrown in parallel without any queue.
2) On completing all requests are still visible in the netstat with "ESTABLISHED" state for next 60 seconds. In case number of such connection is piled up to ~400-500, then new connections are not opened with "2032 (Stream IOError)" exception. Waiting for some time (60 seconds) can help and new connections are available.

What we check and do:
1) Call close() on URLLoader/URLStream. No help.
2) Running of GC - System.gc() and System.pauseForGCIfCollectionImminent(). No help.
3) Reuse of the URLLoader, so there is pool of them to not create new one. No help.

So there are next questions:
1) Is this difference by design or it is possible to change such behavior with some configurational change in application?
2) Is there any way on MacOS to make connection closed or reused to avoid piling up of active connections?

Thanks!

Regards,
Sergey Kobylkin,
Software Technologies LLC

**Comment by ajwfrost** [comment]
Hi

From a bit of analysis and testing here:
1. the requests to open a stream are just passed directly to the operating system, there is no specific queuing or limitation on the AIR side
2. Windows will have a limit for the maximum number of concurrent connections, which is built into their wininet library (I believe it can actually be configured via the registry, I'm only seeing 2 parallel connections on a Win10 machine here)
3. MacOS doesn't appear to have any limit on the number of parallel connections, but yes we're also seeing connection errors when we try opening a large number of requests simultaneously

Trying to add throttling into the runtime itself to avoid opening multiple connections would be a challenge, due to the way it's designed, but it should be feasible to have something as a higher level API that provides this functionality i.e. limiting the number of open connections.

For example, attached is a very hastily-put-together class "ThrottledUrlStream.as" with a test "Main.as" class which uses it, limiting us to no more than 6 open streams at once. If you change the call in Main.as line 29, from `new ThrottledUrlStream` to `new UrlStream`, then you can switch between default behaviour and this utility class.

FYI when testing this with Google, we're then hitting http code 429 errors i.e. the server has decided we're sending too many requests and is just rejecting them. So you may also need to slow down the requests, which could be done by adding timeouts into the ThrottledUrlStream class. But ultimately the success of these calls may well depend on the server behaviour, so not much we can do about those..

thanks
[src.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8618860/src.zip)

**Reply by sergeysoftech** [reply]
Hi,

The analisys and testing section shed some light on internal mechanics, thanks.

About sample (on Mac): it works correct on NUM_STREAMS=400, but URLStream/URLLoader code works correctly on such numbers too. When I raised limit to 1000 (actually 600 will be enough - 512 seems like threshold, but to show the point and avoid any timing/slowness issues let it be 1K), then I see massive "Error: 2032, Stream error" in onError without any onStatus/onComplete. This behavior is the same for ThrottledUrlStream or URLStream.
On testing I've used the local server to avoid any anti-spam systems.

The same sample rewritten on C++ works correctly.

Problem is not the count of connection made by application - creation of connection pool like ThrottledUrlStream in this case is expected and truly to be said already exist in our code. Problem is presence of "already completed" connection in the list of opened connections, like they are not properly closed/disposed by client code. In the initial message I've put a little vague description of this in: 
"On completing all requests are still visible in the netstat with "ESTABLISHED" state for next 60 seconds." 

So from the point of ThrottledUrlStream there are only 5 active connections and new connection can be made. And from the point of the system there are 512 "still active" connections from previous runs of ThrottledUrlStream and new one generates error. Of course in case it happens in 60 seconds timeframe.

Thanks again for the answer!

Regards,
Sergey Kobylkin,
Software Technologies LLC

**Reply by ajwfrost** [reply]
Ah ... I see:
> Problem is presence of "already completed" connection in the list of opened connections, like they are not properly closed/disposed by client code. In the initial message I've put a little vague description of this in:
> "On completing all requests are still visible in the netstat with "ESTABLISHED" state for next 60 seconds."

So we can check on this part and see what's happening with the clean-up... I had expected the connection to be torn down properly once the response is back, regardless of whether that's success or error...

thanks

**Reply by lwmirkk** [reply]
> On completing all requests are still visible in the netstat with "ESTABLISHED" state for next 60 seconds. In case number of such connection is piled up to ~400-500, then new connections are not opened with "2032 (Stream IOError)" exception. Waiting for some time (60 seconds) can help and new connections are available.

ESTABLISHED connections with a 60 seconds timeout is handled by the server, no? I use Litespeed webserver and I control this in server side with litespeed. This is the time a TCP connection stay IDLE waiting for the same "IP" connection to exchange information / data or closed when the time is reached and this "IP" need to make a new TCP request. The name of this config is : Keep-Alive Timeout

If AIR respect or no this KeepAlive timeout, I don't know. '-'

**Reply by vitalysoftech** [reply]
> Ah ... I see:
> 
> > Problem is presence of "already completed" connection in the list of opened connections, like they are not properly closed/disposed by client code. In the initial message I've put a little vague description of this in:
> > "On completing all requests are still visible in the netstat with "ESTABLISHED" state for next 60 seconds."
> 
> So we can check on this part and see what's happening with the clean-up... I had expected the connection to be torn down properly once the response is back, regardless of whether that's success or error...
> 
> thanks

Hi Andrew,

did you have a chance to check it in the code? This is a critical problem for us and we need to get it fixed ASAP. Our macOS application renders tile map and creates URLLoader for each tile. Active panning of map leads to the issue pretty soon. I assume that connections can be reused for requests to the same server, I am not sure whether the logic is implemented in AIR. But the simplest solution would be to close connection after request is done. 

Regards,
Vitaly Shlyaga
Software Technologies LLC

**Reply by ajwfrost** [reply]
Hi

I've just checked the status of this one. So, if we use a webserver that's got the keep-alive, then currently we can see the connection go to "ESTABLISHED" and stay there for quite some time. When we quit the app, it goes to "TIME_WAIT" and stays there for quite a while too.

We've made a change to try to clean up the session which means that we're seeing the "ESTABLISHED" drop back to "TIME_WAIT" pretty much immediately, and this then drops after a few seconds. So I'm not 100% sure whether this is going to be sufficient for what you're after - ideally of course we likely should be trying to cache the connection and reuse it - but this is a bit simpler.

Would it be possible for us to send you a test build to try out?

thanks

**Reply by vitalysoftech** [reply]
Andrew

yes, please. We would like to try it and see how it goes in our case.

Regards,
Vitaly Shlyaga
Software Technologies LLC

**Reply by ajwfrost** [reply]
Great, can you try from the below?

https://transfer.harman.com/message/YFKU2Z6Z3pPhT7Pnv9NAjm

thanks

**Reply by lwmirkk** [reply]
> Hi
> 
> I've just checked the status of this one. So, if we use a webserver that's got the keep-alive, then currently we can see the connection go to "ESTABLISHED" and stay there for quite some time. When we quit the app, it goes to "TIME_WAIT" and stays there for quite a while too.
> 
> We've made a change to try to clean up the session which means that we're seeing the "ESTABLISHED" drop back to "TIME_WAIT" pretty much immediately, and this then drops after a few seconds. So I'm not 100% sure whether this is going to be sufficient for what you're after - ideally of course we likely should be trying to cache the connection and reuse it - but this is a bit simpler.
> 
> Would it be possible for us to send you a test build to try out?
> 
> thanks

I could be wrong, but the change you're making is for the AIR to stop respecting Keep-Alive by the server? That what I understood?
If it is, don't do that, keep allowing the AIR to respect KeepAlive, this is very important, that way the app doesn't close the connection if it's not to close. Or the AIR then has the option to accept or not to respect KeepAlive, but by default to respect.....

For example, today we use the app with thousands of active TCP / HTTP connections and we need to have KeepAlive for a few seconds to facilitate the use the same route as the CDN and Internap/Nordic. On the server we configure KeepAlive for the time we need, if the connection closes too soon without KeepAlive, the next request will make a new TCP connection on the server and the CDN and Internap will do a new route check for this connection, being overkill for us. With KeepAlive we ensure that our app has an almost constant connection on the best route using the same TCP connection that is still open / keepalive. and TIME_WAIT is when the connection is being closed, if it's in TIME_WAIT it doesn't mean it's in KeepAlive, they've already been closed, waiting to be garbage collected or something.

Well, I don't know what they intend to do, but I may have misunderstood, but I hope that AIR continues to respect the server's KeepAlive. Thanks.

**Reply by ajwfrost** [reply]
Okay, yes so you're right, essentially the change we'd made there was to try to tear down the connection so that we don't end up in a situation where there are too many open connections hence causing the failures as mentioned initially.

And yes, this isn't really a proper solution, the ideal one would be to properly cache and reuse the connection if it's still open - which should then fit with your requirement here. Let me go back to the team and check how much effort it might be to do this, currently I believe there's just a new URL session being kicked off for every AS3 URL request...

thanks

**Reply by vitalysoftech** [reply]
Andrew, lwmirkk,

doesn't the fact that AIR creates new URL session each time we create an URL request mean that it doesn't reuse open connections? So the keep-alive logic is already not supported by macOS AIR runtime.

Regards,
Vitaly Shlyaga
Software Technology LLC

**Reply by lwmirkk** [reply]
To be honest, I don't know how the AIR/OS part of the TCP connection works today. I didn't get to track these details in depth, but what we've tested and used is that if we take the Keep Alive off the server or its value is lower than the next request made by AIR, a new TCP connection will be opened. When we put KeepAlive and make a new request before the KeepAlive time expires, the server continues with this same TCP connection, in our case it is better to have thousands of TCP connections open, as we are sure that there will be a new request before the KeepAlive expires. So we made the server have a KeepAlive a few seconds longer than we know our app "pings" the server.
So if we know that the app has a request every 10 seconds, today we have KeepAlive of 15 seconds. So en most of cases the new request will always use the same TCP. Helping the server overload and as we use Internap and Nordic and CDN, this help the re-use of same best route according with our extended tests.

In this same example, when we put KeepAlive below 10 seconds, every new request opened a new TCP.

In a test with 700 TCP connections, when KeepAlive was less than the time between each app request, there were about 250 ESTABLISHED connections and 6000 in TIME_WAIT, as they were already closed...
When we configure KeepAlive to be longer than the time of one request and another, the almost 700 connections were ESTABLISHED and only 300 in TIME_WAIT for example, in our case, we know that as we are sure of these constant requests, KeepAlive saves us. But if we didn't use the app, if it was a normal website, then certainly KeepAlive could be smaller.

Today we only use the app on Windows and Android, we don't know how it behaves on macOS and if the macOS system interferes with these connections differently.

When I say, thousands of requests we make, I didn't explain it right, the thousands of constant connections are from simultaneous apps. There are no thousands of **concurrent** connections in the same app. In the same app we try to handle a queue of requests, so I believe it's no more than 20 different requests simultaneously in the same app.

@vitalysoftech 
Well, that's how our app works today, now if I needed to make hundreds of simultaneous TCP requests in the same app, I believe this will have a limit by the system itself, right? Each system has a different internal limit, doesn't it?
And even if I need to do that, I believe that since AIR works today and as we've already tested, AIR seems to respect the server, so if you tweak KeepAlive on the server, it doesn't work? Our TCP KeepAlive settings were done on LiteSpeed ​​servers.
When we adjust it to 0 per example (disable keepalive), the connection is dropped to TIME_WAIT as soon its finished / closed.
And when we set to what we need, the connection keep ESTABLISHED in server TCP list.

This is the explanation that is in the Litespeed configuration:
![image](https://user-images.githubusercontent.com/36115546/169528438-3f60c035-679f-48bc-a4bb-499dba24d507.png)


P.S.:
The PHP KeepAlive is different from Server KeepAlive, this adjust is in the server config, to KeepAlive the TCP request, the default one in Litespeed is 2 if I remember.
PHP KeepAlive (if you use PHP or Rails or something similar) for us, this config is 0. Because this keepalive if different from the server TCP keepalive.

P.S - 2: 
this keepalive is not just a return header from the server, it's an internal keepalive from the server, the app doesn't know this, it will make the new request, and the server will decide if it needs to use this "open" TCP or make a new one for this one client.

**Reply by vitalysoftech** [reply]
@lwmirkk,
We don't control server's keep-alive setting, because this is 3rd party service. Looks like you don't experience the issue, because you don't use macOS build, but only Android and Windows. On Windows our app also behaves correctly. I guess that each platform has its own network stack implementation under the AIR hood. From what we see now, macOS implementation has a bug. When we send request AIR always opens connection that will be used only for this request. When request is done, connection is left in memory for 1 minute until it is finally closed (I guess by server, which has 1 minute timeout for keep alive). For example, if you do 2 requests to the same server in a series, the second request won't reuse connection of the first request - it will create new one. So sending many requests during a minute can lead to the case, when new connection cannot be opened, because of a limit.

**Reply by lwmirkk** [reply]
> Looks like you don't experience the issue, because you don't use macOS build, but only Android and Windows. On Windows our app also behaves correctly. I guess that each platform has its own network stack implementation under the AIR hood

I think you are correct about this.

> We don't control server's keep-alive setting, because this is 3rd party service. 
This is not good, because per I searched, some server use up to 2 hours keepalive requests.

Probably Andrew and the team need to test the thinks on MacOS to try find the difference between MacOS and Windows in this case.

If you (@vitalysoftech and @ajwfrost) need to test ours current keepalive - 15 seconds requests and see if this make something different in a request in mac with our server, I can send you a link, so you can make requests. If you need to test some specific static file(s), just send me they and I will put them in a link in oour server, so you can make the test in macOS. =)

**Reply by ajwfrost** [reply]
There may be differences here from the server vs the client but from my understanding (and having played a little with different servers) the connection that the macOS machine has (created via `NSURLSession`) will either stay open for a while (if there's a keep-alive request) or will close down fairly quickly (if there's not or if there's a connection: close request).

But the issue here (I think!) is when a single app is repeatedly requesting data from a server, e.g. per your example, every 10 seconds it fires off an HTTP request. If the keep-alive is set to 15 seconds, then all of these requests should be sent over a persistent network connection, i.e. there should only be a single 'established' connection listed via netstat.

So from a bit of research here, the issue may simply be that in AIR we are creating a new `NSURLSession` object for each URL request. It seems that normally, `NSURLSession` would internally track the connections and decide whether to reuse one vs opening a new one, but if we create a new session then this re-use doesn't happen.

We can look at how much effort it would be to switch to a system where we have a single session and then just kick off the individual data requests, it's going to need a bit of a redesign I think....

thanks

**Reply by vitalysoftech** [reply]
@ajwfrost, typical solution is to have a connection pool. When session/request is created it requests a connection from the pool which either issues existing one that is still open or creates a new one. When request is done - connection is returned to the pool. This is how it is implemented on some platforms, e.g. .NET. BTW, I think, that you can compare macOS and Windows URLRequest implementations and API they use under the hood to realize why the problem is only on macOS. May be macOS doesn't have this connection pool behind API they expose, so it is necessary to implement it manually on AIR side.  

Meantime we are testing the custom build of AIR runtime to see whether it became better.

Thanks!

**Reply by vitalysoftech** [reply]
@ajwfrost, we checked the custom build of AIR SDK and cannot reproduce the issue any longer. I realize that if we put the application under higher stress we may have a chance to reproduce it. But what concerns me now is that macOS version doesn't reuse connections (I assume it didn't reuse from very beginning) and that can lead to slower work of application than it could be with proper reusing, especially in cases like in ours, when we send many requests to the same server.

Did you have a chance to estimate effort required for correct fix of the problem?

Thanks!

Regards,
Vitaly

**Reply by ajwfrost** [reply]
Hi @vitalysoftech  - apologies for the delay coming back to you on this. You're right in that we don't appear to have ever been reusing connections, from what I can see it was always setting up a new NSURLSession for every connection -> which is bad practice, we should actually just have a single session and let that then deal with all the connection handling.

We had a bit of fun with delegates (this would get a lot easier if we could limit our support to Monterey!) but it looks like we've got a version that uses a shared session whilst still being able to identify the appropriate AS3 url requests from each of the delegate callbacks... (at least, using our test code, which is a bit basic, we're getting 4000 URL requests completing successfully, and only one actual connection according to the netstat information - these are all to the same location)

I'll look at getting another test runtime over to you for you to check on this, but am hopeful that this will make things a bit better!

thanks

**Reply by vitalysoftech** [reply]
Hi @ajwfrost ! Good to hear and will be glad to test on our end. BTW, our minimum system requirements for macOS is 10.15 (Catalina). Will solution work on it?

Regards,
Vitaly

**Reply by ajwfrost** [reply]
Yes should be fine from 10.12 upwards (and that's just a limitation because of a quirk of the signatures, if you un-sign things then it should work on 10.10+). 

You can grab the test runtime from:
https://transfer.harman.com/message/7ZPSHjjHTYoV5CE59p7Yjz

thanks

**Reply by vitalysoftech** [reply]
Hi , @ajwfrost. Unfortunately, the application crashes with the new runtime. Here is the crash log: 

https://gist.github.com/vitalysoftech/0235d8bc35067483b16041cfba5b7888

Thanks.

**Reply by ajwfrost** [reply]
Hmmm... annoyingly, I don't think we have the symbols from that test build, but it wasn't crashing in our test app.
Is there any chance we can get the AIR application that you're using to reproduce this?
thanks

**Reply by vitalysoftech** [reply]
@ajwfrost, you can download the application from here: https://www.mapbusinessonline.com/App-Download.aspx (button to download macOS version is below Windows).

**Reply by ajwfrost** [reply]
Thanks @vitalysoftech  -> we managed to reproduce it and fix it, it was relating to https authentication...

**Reply by vitalysoftech** [reply]
Thanks, @ajwfrost . Will you send us updated runtime?

**Reply by ajwfrost** [reply]
Sure, I can add another link up here later on today..

**Reply by ajwfrost** [reply]
Sorry for the delay! Please try with https://transfer.harman.com/message/EqHzLF6sBmUofK8Szpk0YN

**Reply by vitalysoftech** [reply]
@ajwfrost, it works good! The same HTTPS2 connection is reused between all requests to the same server.

**Reply by ajwfrost** [reply]
Great, thanks for confirming.. I was kind of hoping so, as this change had been merged into our release candidate that's going through QA and should get the green light for a release on Monday!

cheers

**Reply by vitalysoftech** [reply]
@ajwfrost, great! We are looking forward!

**Reply by vitalysoftech** [reply]
@ajwfrost , I see that there are notes published for new release, but the latest version for download is still previous one (33.1.1.856). Is deployment still in progress?

**Reply by ajwfrost** [reply]
@vitalysoftech just an oversight, sorry! updated now...
thanks
